stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

```diff
@@ -64,12 +67,74 @@ func posixPipe() -> (readFD: CInt, writeFD: CInt) {
 /// stderr.
 public func spawnChild(_ args: [String])
   -> (pid: pid_t, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  // The stdout, stdin, and stderr from the child process will be redirected
+  // to these pipes.
+  let childStdout = posixPipe()
+  let childStdin = posixPipe()
+  let childStderr = posixPipe()
+
+#if os(Android)
+  // posix_spawn isn't available on Android. Instead, we fork and exec.
+  // To correctly communicate the exit status of the child process to this
+  // (parent) process, we'll use this pipe.
+  let childToParentPipe = posixPipe()
+
+  let pid = fork()
+  precondition(pid >= 0, "fork() failed")
+  if pid == 0 {
+    // pid of 0 means we are now in the child process.
+    // Capture the output before executing the program.
+    dup2(childStdout.writeFD, STDOUT_FILENO)
+    dup2(childStdin.readFD, STDIN_FILENO)
+    dup2(childStderr.writeFD, STDERR_FILENO)
+
+    // Set the "close on exec" flag on the parent write pipe. This will
+    // close the pipe if the execve() below successfully executes a child
+    // process.
+    let closeResult = fcntl(childToParentPipe.writeFD, F_SETFD, FD_CLOEXEC)
+    let closeErrno = errno
+    precondition(
+      closeResult == 0,
+      "Could not set the close behavior of the child-to-parent pipe; " +
+      "errno: \(closeErrno)")
+
+    // Start the executable. If execve() does not encounter an error, the
+    // code after this block will never be executed, and the parent write pipe
+    // will be closed.
+    withArrayOfCStrings([Process.arguments[0]] + args) {
+      execve(Process.arguments[0], $0, _getEnviron())
+    }
+
+    // If execve() encountered an error, we write the errno encountered to the
+    // parent write pipe.
+    let errnoSize = sizeof(errno.dynamicType)
+    var execveErrno = errno
+    let writtenBytes = withUnsafePointer(&execveErrno) {
+      write(childToParentPipe.writeFD, UnsafePointer($0), errnoSize)
+    }
+
+    let writeErrno = errno
+    if writtenBytes > 0 && writtenBytes < errnoSize {
+      // We were able to write some of our error, but not all of it.
+      // FIXME: Retry in this case.
+      preconditionFailure("Unable to write entire error to child-to-parent " +
+                          "pipe.")
+    } else if writtenBytes == 0 {
+      preconditionFailure("Unable to write error to child-to-parent pipe.")
+    } else if writtenBytes < 0 {
+      preconditionFailure("An error occurred when writing error to " +
+                          "child-to-parent pipe; errno: \(writeErrno)")
+    }
+
+    // Close the pipe when we're done writing the error.
+    close(childToParentPipe.writeFD)
```

Comment:
@modocache I'm currently looking from my phone so it's likely that it's there and I don't see it. After the fork, the parent and the child have both opened the read _and_ the writeFD of the pipe. It's important (before doing anything else) to close the writeFD in the parent and the readFD in the child.

Because what we're looking for is a unidirectional comm channel from child to parent. The close will only be seen by the parent when _the last writer_ closes the pipe. If writeFD is also open in the parent, the child closing the writeFD is not the last writer.

Code should be

    if pid == 0 {
        close(pipe.readFD)
        fcntl(pipe.writeFD, close_on_exec)
        execve()
        Write(pipe.writeFD, errno
         Close(pipe.writeFD
    } else if pid >= 0 {
        Close(pipe.writeFD) // very important 
        bytes = read(pipe.readFD, buffer)
        analyse bytes and buffer...
    }

Excuse the horrible pseudo code, not fun to type on phone.

Does that make sense?